controller/tests/controller_server/csi_controller_server_test.py

```diff
@@ -2,19 +2,242 @@
+        self.assertTrue("secret" in context.details)
+
+        self.request.secrets = {"username": "user", "management_address": "mg"}
+        self.get_create_object_method()(self.request, context)
+        self.assertEqual(context.code, grpc.StatusCode.INVALID_ARGUMENT, "password is missing in secrets")
+        self.assertTrue("secret" in context.details)
+
+        self.request.secrets = {"username": "user", "password": "pass"}
+        self.get_create_object_method()(self.request, context)
+        self.assertEqual(context.code, grpc.StatusCode.INVALID_ARGUMENT, "mgmt address is missing in secrets")
+        self.assertTrue("secret" in context.details)
+
+        self.request.secrets = []
+
+    def _test_create_object_with_array_connection_exception(self, a_enter):
+        a_enter.side_effect = [Exception("error")]
+        context = utils.FakeContext()
+        self.get_create_object_method()(self.request, context)
+        self.assertEqual(context.code, grpc.StatusCode.INTERNAL, "connection error occured in array_connection")
+        self.assertTrue("error" in context.details)
+
+    def _test_create_object_with_get_array_type_exception(self, a_enter, array_type):
+        a_enter.return_value = self.mediator
+        context = utils.FakeContext()
+        array_type.side_effect = [array_errors.FailedToFindStorageSystemType("endpoint")]
+        self.get_create_object_method()(self.request, context)
+        self.assertEqual(context.code, grpc.StatusCode.INTERNAL, "failed to find storage system")
+        msg = array_errors.FailedToFindStorageSystemType("endpoint").message
+        self.assertTrue(msg in context.details)
+
+
+class TestControllerServerCreateSnapshot(AbstractControllerTest):
+
+    def get_create_object_method(self):
+        return self.servicer.CreateSnapshot
+
+    def get_create_object_method_response(self):
+        return csi_pb2.CreateSnapshotResponse
+
+    @patch("controller.array_action.array_mediator_xiv.XIVArrayMediator._connect")
+    def setUp(self, connect):
+        self.fqdn = "fqdn"
+        self.mediator = XIVArrayMediator("user", "password", self.fqdn)
+        self.mediator.client = Mock()
+        self.mediator.get_snapshot = Mock()
+        self.mediator.get_snapshot.return_value = None
+
+        self.servicer = ControllerServicer(self.fqdn)
+
+        self.request = Mock()
+        self.request.secrets = {"username": "user", "password": "pass", "management_address": "mg"}
+        self.request.parameters = {}
+        self.capacity_bytes = 10
+        self.request.name = snap_name
+        self.request.source_volume_id = "A9000:12345678"
+
+    @patch("controller.array_action.array_connection_manager.ArrayConnectionManager.__enter__")
+    @patch("controller.array_action.array_connection_manager.ArrayConnectionManager.__exit__")
+    def test_create_snapshot_with_empty_name(self, a_enter, a_exit):
+        self._test_create_object_with_empty_name(a_enter)

```

Comment:
I thought the order of patch is [reversed](https://docs.python.org/3/library/unittest.mock.html#quick-guide) to the parameters order. how does it work? is this a false positive?

Reply:
It was copy paste from volume test.

Reply:
we shouldn't use their buggy patch code for new tests.

I think it will be easier for you to fix the volume tests before using them (because the fix should be to remove the unnecessary-yet-required parameter),
but if not, then at least write your patches in the correct order

Reply:
I started to remove what seems to be unused and the tests failed.

Reply:
Fixed order/removed unused where possible